### PR TITLE
bpo-30522: Implemented a method to allow setting a logging.StreamHand…

### PR DESCRIPTION
**Pull request by vsajip**
…ler's stream.

https://bugs.python.org/issue30522

### CODE REVIEW

File: Doc/library/logging.handlers.rst [hunk 1]
```diff
@@ -59,13 +59,25 @@ and :meth:`flush` methods).
       :meth:`close` method is inherited from :class:`~logging.Handler` and so
       does no output, so an explicit :meth:`flush` call may be needed at times.
 
+   .. method:: setStream(stream)
+
+      Sets the instance's stream to the specified value, if it is different.
+
+      :param stream: The stream that the handler should use.
```
Comment (pitrou) [inline Doc/library/logging.handlers.rst]: Do we support the `:param:` markup? I don't think I've seen it before in our docs.

File: Doc/library/logging.handlers.rst [hunk 2]
```diff
@@ -59,13 +59,25 @@ and :meth:`flush` methods).
       :meth:`close` method is inherited from :class:`~logging.Handler` and so
       does no output, so an explicit :meth:`flush` call may be needed at times.
 
+   .. method:: setStream(stream)
```
Comment (pitrou) [inline Doc/library/logging.handlers.rst:62]: This method needs a `versionadded` marker.

File: Doc/library/logging.handlers.rst [hunk 3]
```diff
@@ -59,13 +59,25 @@ and :meth:`flush` methods).
...
+      :return: the old stream, if the stream was changed, or *None* if it wasn't.
+
 .. versionchanged:: 3.2
    The ``StreamHandler`` class now has a ``terminator`` attribute, default
    value ``'\n'``, which is used as the terminator when writing a formatted
    record to a stream. If you don't want this newline termination, you can
    set the handler instance's ``terminator`` attribute to the empty string.
    In earlier versions, the terminator was hardcoded as ``'\n'``.
 
+.. versionchanged:: 3.7
```
Comment (pitrou) [inline Doc/library/logging.handlers.rst]: This is better replaced with a `versionadded` marker for the method itself, see above.

**Review by pitrou — COMMENTED** [review]
You probably also want to a NEWS blurb here.

File: Doc/library/logging.handlers.rst (hunk 1 above)
Comment (vsajip) [reply Doc/library/logging.handlers.rst]: AFAIK it's part of vanilla Sphinx. The docs built OK with it in.

File: Lib/logging/__init__.py [hunk 4]
```diff
@@ -997,6 +997,26 @@ def emit(self, record):
...
+        Returns the old stream, if the stream was changed, or None
+        if it wasn't.
+        """
+        if stream is self.stream:
+            result = None
+        else:
+            result = self.stream
+            self.acquire()
+            try:
+                self.flush()
```
Comment (qingyunha) [inline Lib/logging/__init__.py:1014]: Is the flush need here? The only place to write steam is emit()，that already has a flush. If the flush is not needed, the lock is also not needed.

Comment (vsajip) [reply Lib/logging/__init__.py:1014]: This may seem a bit defensive but is unlikely to be in performance-critical code, I would have thought.

**Pull request merged**